lib/lte_link_control/lte_lc.c

```diff
@@ -773,1088 +57,183 @@ int lte_lc_deregister_handler(lte_lc_evt_handler_t handler)
-		"%hu," /* <loop> */
-		"%hu," /* <return_to_pattern> */
-		"%hu," /* <band_optimization> */
-		"\"%40[^\"]\","  /* <pattern_1> */
-		"\"%40[^\"]\","  /* <pattern_2> */
-		"\"%40[^\"]\","  /* <pattern_3> */
-		"\"%40[^\"]\"",  /* <pattern_4> */
-		&loop_tmp, &cfg->return_to_pattern, &cfg->band_optimization,
-		pattern_buf[0], pattern_buf[1], pattern_buf[2], pattern_buf[3]
-	);
-	if (err == -NRF_EBADMSG) {
-		return -ENOENT;
-	} else if (err < 0) {
-		return -EFAULT;
-	} else if (err < 4) {
-		/* Not all parameters and at least one pattern found */
-		return -EBADMSG;
-	}
-
-	cfg->loop = loop_tmp;
-
-	/* Pattern count is matched parameters minus 3 for loop, return_to_pattern
-	 * and band_optimization.
-	 */
-	cfg->pattern_count = err - 3;
-
-	LOG_DBG("Pattern 1: %s", pattern_buf[0]);
-
-	err = parse_periodic_search_pattern(pattern_buf[0], &cfg->patterns[0]);
-	if (err) {
-		LOG_ERR("Failed to parse periodic search pattern");
-		return err;
-	}
-
-	if (cfg->pattern_count >= 2) {
-		LOG_DBG("Pattern 2: %s", pattern_buf[1]);
-
-		err = parse_periodic_search_pattern(pattern_buf[1], &cfg->patterns[1]);
-		if (err) {
-			LOG_ERR("Failed to parse periodic search pattern");
-			return err;
-		}
-	}
-
-	if (cfg->pattern_count >= 3) {
-		LOG_DBG("Pattern 3: %s", pattern_buf[2]);
-
-		err = parse_periodic_search_pattern(pattern_buf[2], &cfg->patterns[2]);
-		if (err) {
-			LOG_ERR("Failed to parse periodic search pattern");
-			return err;
-		}
-	}
-
-	if (cfg->pattern_count == 4) {
-		LOG_DBG("Pattern 4: %s", pattern_buf[3]);
-
-		err = parse_periodic_search_pattern(pattern_buf[3], &cfg->patterns[3]);
-		if (err) {
-			LOG_ERR("Failed to parse periodic search pattern");
-			return err;
-		}
-	}
-
-	return 0;
+	return periodicsearchconf_get(cfg);
 }
 
 int lte_lc_reduced_mobility_get(enum lte_lc_reduced_mobility_mode *mode)
 {
-	int ret;
-	uint16_t mode_tmp;
-
-	if (mode == NULL) {
-		return -EINVAL;
-	}
-
-	ret = nrf_modem_at_scanf("AT%REDMOB?", "%%REDMOB: %hu", &mode_tmp);
-	if (ret != 1) {
-		LOG_ERR("AT command failed, nrf_modem_at_scanf() returned error: %d", ret);
-		return -EFAULT;
-	}
-
-	*mode = mode_tmp;
-
-	return 0;
+	return redmob_get(mode);
 }
 
 int lte_lc_reduced_mobility_set(enum lte_lc_reduced_mobility_mode mode)
 {
-	int ret = nrf_modem_at_printf("AT%%REDMOB=%d", mode);
-
-	if (ret) {
-		/* Failure to send the AT command. */
-		LOG_ERR("AT command failed, returned error code: %d", ret);
-		return -EFAULT;
-	}
-
-	return 0;
+	return redmob_set(mode);
 }
 
 int lte_lc_factory_reset(enum lte_lc_factory_reset_type type)
 {
-	return nrf_modem_at_printf("AT%%XFACTORYRESET=%d", type) ? -EFAULT : 0;
+	return xfactoryreset_reset(type);
 }
 
 static int lte_lc_sys_init(void)
 {
-	struct k_work_queue_config cfg = {
-		.name = "lte_lc_work_q",
-	};
-
-	k_work_queue_start(
-		&lte_lc_work_q,
-		lte_lc_work_q_stack,
-		K_THREAD_STACK_SIZEOF(lte_lc_work_q_stack),
-		K_LOWEST_APPLICATION_THREAD_PRIO,
-		&cfg);
+	work_q_start();
```

Comment:
Its a bit pointless to have a function that only calls one void-function.

Can you just put work_q_start into SYS_INIT() and drop the lte_lc_sys_init()?

Reply:
That wouldn't work because `SYS_INIT` expects a function that returns `int`. On top of that, keeping this function is not to hurt much neither the code nor the performance. If in the future functionalities need to be added to init, they can be added there instead of refactoring the work queue start out of the `SYS_INIT` again.